src/pg/pg-store.ts

```diff
@@ -0,0 +1,513 @@
+  }
+
+  async rollBackInscriptionTransfer(args: { genesis_id: string; output: string }): Promise<void> {
+    let inscription_id: number | undefined;
+    await this.sqlWriteTransaction(async sql => {
+      const inscription = await sql<{ id: number }[]>`
+        SELECT id FROM inscriptions WHERE genesis_id = ${args.genesis_id}
+      `;
+      if (inscription.count === 0) {
+        logger.warn(args, `PgStore ignoring rollback for a transfer that does not exist`);
+        return;
+      }
+      inscription_id = inscription[0].id;
+      await sql`
+        DELETE FROM locations
+        WHERE inscription_id = ${inscription_id} AND output = ${args.output}
+      `;
+    });
+    if (inscription_id) await this.normalizeInscriptionLocations({ inscription_id });
+  }
+
+  private async normalizeInscriptionLocations(args: { inscription_id: number }): Promise<void> {
+    await this.sql`
+      WITH i_genesis AS (
+        SELECT id FROM locations
+        WHERE inscription_id = ${args.inscription_id}
+        ORDER BY block_height ASC
+        LIMIT 1
+      ), i_current AS (
+        SELECT id FROM locations
+        WHERE inscription_id = ${args.inscription_id}
+        ORDER BY block_height DESC
+        LIMIT 1
+      )
+      UPDATE locations SET
+        current = (CASE WHEN id = (SELECT id FROM i_current) THEN TRUE ELSE FALSE END),
+        genesis = (CASE WHEN id = (SELECT id FROM i_genesis) THEN TRUE ELSE FALSE END)
+      WHERE inscription_id = ${args.inscription_id}
+    `;
+  }
+
+  async getInscriptionCurrentLocation(args: { output: string }): Promise<DbLocation | undefined> {
+    const result = await this.sql<DbLocation[]>`
+      SELECT ${this.sql(LOCATIONS_COLUMNS)}
+      FROM locations
+      WHERE output = ${args.output}
+      AND current = TRUE
+    `;
+    if (result.count === 0) {
+      return undefined;
+    }
+    return result[0];
+  }
+
+  async getInscriptionContent(
+    args: InscriptionIdentifier
+  ): Promise<DbInscriptionContent | undefined> {
+    const result = await this.sql<DbInscriptionContent[]>`
+      SELECT content, content_type, content_length
+      FROM inscriptions
+      WHERE ${
+        'genesis_id' in args
+          ? this.sql`genesis_id = ${args.genesis_id}`
+          : this.sql`number = ${args.number}`
+      }
+    `;
+    if (result.count > 0) {
+      return result[0];
+    }
+  }
+
+  async getInscriptionETag(args: InscriptionIdentifier): Promise<string | undefined> {
+    const result = await this.sql<{ etag: string }[]>`
+      SELECT date_part('epoch', l.timestamp)::text AS etag
+      FROM locations AS l
+      INNER JOIN inscriptions AS i ON l.inscription_id = i.id
+      WHERE ${
+        'genesis_id' in args
+          ? this.sql`i.genesis_id = ${args.genesis_id}`
+          : this.sql`i.number = ${args.number}`
+      }
+      AND l.current = TRUE
+    `;
+    if (result.count > 0) {
+      return result[0].etag;
```

Comment:
If these etags are all time based, you could possibly implement very efficient caching with `cache-control: max-age=x` -- the browser could reliably use its cached data (or cloudflare's) without having to revalidate against the API. You'd want to only do this when the datetime is old enough to avoid bitcoin reorgs.

Not suggesting this PR should do that, could be something addressed later on.

Reply:
Great idea, opened https://github.com/hirosystems/ordinals-api/issues/53